File: clojure/jherrlin/day3/problem2.clj
(ns day-3-problem-2
  (:require [day-3-problem-1 :as problem1]))

(defn determine-value-to-save [pred-fn frequencies']
  (let [zeroes (get frequencies' 0 0)
        ones   (get frequencies' 1 0)]
    (pred-fn zeroes ones)))

(defn calculate-most-commons [pred-fn parsed-data]
  (->> parsed-data
       problem1/rotate-vector
       (map (juxt frequencies identity))
       (map-indexed vector)
       (map (fn [[idx [freq coll]]]
              {:column-id   idx
               :frequencies freq
               :coll        coll
               :value-to-save (determine-value-to-save pred-fn freq)}))))

(defn determine-oxygen-generator-rating [select-value-to-save-fn diagnostic-vector]
  (loop [parsed' diagnostic-vector
         counter 0]
    (if (= 1 (count parsed'))
      parsed'
      (let [commons (->> parsed'
                         (calculate-most-commons select-value-to-save-fn)
                         (map (juxt :column-id identity))
                         (into {}))]
        (recur
         (filter (fn [row] (= (nth row counter)
                              (get-in commons [counter :value-to-save]))) parsed')
         (inc counter))))))

(defn determine-value-to-save-oxygen [zeroes ones]
  (if (>= ones zeroes) 1 0))

(defn determine-value-to-save-CO2 [zeroes ones]
  (if (<= zeroes ones) 0 1))

(defn life-support-rating [input]
  (*
   (->> (problem1/parse-input input)
        (determine-oxygen-generator-rating determine-value-to-save-oxygen)
        first
        (problem1/binary-coll-to-decimal))
   (->> (problem1/parse-input input)
        (determine-oxygen-generator-rating determine-value-to-save-CO2)
        first
        (problem1/binary-coll-to-decimal))))

;; test input
(life-support-rating problem1/test-input)
;; => 230

;; real input
(life-support-rating (slurp "./input.txt"))
;; => 7928162

File: python/jherrlin/day3/problem2.py
import collections

input = """00100
11110
10110
10111
10101
01111
00111
11100
10000
11001
00010
01010"""

test_matrix = [[int(y) for y in list(x)] for x in input.splitlines()]

with open('input.txt', 'r') as f:
    matrix = [[int(y) for y in list(x.strip())] for x in f.readlines()]

def find(matrix, oxygen_or_co2, idx=0):
    if len(matrix) == 1:
        return int("".join([str(x) for x in matrix[0]]), 2)

    frequencies = collections.Counter([x[idx] for x in matrix])
    most_common = oxygen_or_co2(frequencies.get(0), frequencies.get(1))
    matrix = list(filter(lambda x: x[idx] == most_common, matrix))
    idx += 1
    return find(matrix, oxygen_or_co2, idx)

# test input
a = find(test_matrix, lambda zeroes, ones: 1 if zeroes <= ones else 0)
b = find(test_matrix, lambda zeroes, ones: 0 if zeroes <= ones else 1)
print(a * b)

# real input
c = find(matrix, lambda zeroes, ones: 1 if zeroes <= ones else 0)
d = find(matrix, lambda zeroes, ones: 0 if zeroes <= ones else 1)
print(c * d)
